compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -928,6 +969,128 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
       }
     }
 
+    // Static forwarders -------------------------------------------------------
+
+    /* This mimics the logic in BCodeHelpers.addForwarders and the code that
+     * calls it, except that we consider as collisions only methods with the
+     * same `MethodName` (including signature) and that are also
+     * `PublicStatic`. Anything else is not a collision as far the IR is
+     * concerned. We need this in some cases, for example the class `Integer`
+     * has a `def hashCode(): Int` and a `static def hashCode(Int): Int`. The
+     * JVM back-end considers them as colliding because they have the same
+     * name, but we must not.
+     */
+
+    /** Is the given Scala class, interface or module class a candidate for
+     *  static forwarders?
+     */
+    def isCandidateForForwarders(sym: Symbol): Boolean = {
+      // it must be a top level class (name contains no $s)
+      !settings.noForwarders && exitingPickler {
+        !sym.name.containsChar('$') && !sym.isNestedClass
+      }
+    }
+
+    /** Adds static forwarders to the members of a class or interface for
+     *  methods of its companion object.
+     *
+     *  This is only done if there exists a companion object and it is not a JS
+     *  type.
+     *
+     *  Precondition: `isCandidateForForwarders(sym)` is true
+     */
+    def addStaticForwardersForClassOrInterface(
+        existingMembers: List[js.MemberDef], sym: Symbol)(
+        implicit pos: Position): List[js.MemberDef] = {
+      val module = sym.companionModule
+      if (module == NoSymbol) {
+        existingMembers
+      } else {
+        val moduleClass = module.moduleClass
+        if (!isJSType(moduleClass))
+          addStaticFowardersFromModuleClass(existingMembers, moduleClass)
+        else
+          existingMembers
+      }
+    }
+
+    private lazy val dontUseExitingUncurryForForwarders =
+      scala.util.Properties.versionNumberString.startsWith("2.11.")
+
+    /** Adds static forwarders for the methods of a module class to a list of
+     *  existing members.
+     *
+     *  Precondition: `isCandidateForForwarders(moduleClass)` is true
+     */
+    def addStaticFowardersFromModuleClass(existingMembers: List[js.MemberDef],
+        moduleClass: Symbol)(
+        implicit pos: Position): List[js.MemberDef] = {
```

Comment:
I find the way this method works a bit strange: Why does it not simply generate the static forwarders and return them? Returning the first argument as well seems odd.